Describe proxy refactoring in the Application Gateway (#7784)
kyma-project/kyma: docs/application-connector/03-08-application-gateway-proxying-requests.md

---
title: Application Gateway
type: Details
---

> >**CAUTION:** This document describes the proxy service in the alternative **gatewayOncePerNamespace** [mode](#architecture-application-connector-components-application-operator). To read about the proxy service in the default legacy mode, see [this](#architecture-application-gateway) document. 

Reviewer: CAUTION or NOTE? 🤔 
https://kyma-project.io/community/guidelines/content/#formatting-formatting-panels

Author: See [here](https://github.com/kyma-project/kyma/pull/7784#discussion_r397877011)

> >**CAUTION:** This document describes the proxy service in the alternative **gatewayOncePerNamespace** [mode](#architecture-application-connector-components-application-operator). To read about the proxy service in the default legacy mode, see [this](#architecture-application-gateway) document. 

Reviewer: ```suggestion
>**CAUTION:** This document describes the proxy service in the alternative [**gatewayOncePerNamespace** mode](#architecture-application-connector-components-application-operator). To read about the proxy service in the default legacy mode, see [this](#architecture-application-gateway) document. 
```

Author: See [this](https://github.com/kyma-project/kyma/pull/7784#discussion_r397880117)

        }
    }
}
```

An example **CONFIGURATION** for APIs secured with BasicAuth looks as follows:

```json
{
    "credentials": {
      "username":"{USERNAME}",
      "password":"{PASSWORD}"
    },
    "csrfConfig": {
        "tokenUrl": "{CSRF_TOKEN_URL}"
    },
    "requestParameters": {
        "headers": {
            "Content-Type": ["application/json"]
        },
        "queryParameters": {
            "limit": ["50"]
        }
    }
}
```

> **NOTE:** All APIs defined in a single Secret use the same configuration - the same credentials, CSRF tokens, and request parameters.

### Calling the Application Gateway
​
The Secret name and the API name are specified as path variables in the following format:

```
http://my-namespace-application-gateway:8080/secret/{SECRET_NAME}/api/{API_NAME}/{CUSTOM_PATH}
```
​
In order for the Application Gateway to properly read Secrets, they must exist in the same Namespace as the Gateway.

## Caching

To ensure optimal performance, the Application Gateway caches the OAuth tokens and CSRF tokens it obtains. If the service doesn't find valid tokens for the call it makes, it gets new tokens from the OAuth server and the CSRF token endpoint.

> Additionally, the service caches ReverseProxy objects used to proxy the requests to the underlying URL.

Reviewer: ```suggestion
Additionally, the service caches ReverseProxy objects used to proxy requests to the underlying URL.
```

---
title: Proxying requests by the Application Gateway
type: Details
---

> The Application Gateway proxies requests from lambda functions and services in Kyma to external APIs. When the Application Gateway works in the alternative `gatewayOncePerNamespace` [mode](#architecture-application-connector-components-application-operator), it proxies the requests based on the configuration stored in Secrets.

Reviewer: `gatewayOncePerNamespace` -> shouldn't we use the **bold** formatting here, just as we do with parameters?

​
### Proxy configuration
​

> The Secret expected by the Application Gateway has the following structure:

Reviewer: ```suggestion
The Application Gateway expects the Secret that has the following structure:
```

```
data:
    {API_NAME_1}_TARGET_URL: {BASE64_ENCODED_API1_URL}
    {API_NAME_2}_TARGET_URL: {BASE64_ENCODED_API2_URL}
    CONFIGURATION: {BASE64_ENCODED_CONFIG_JSON}
    CREDENTIALS_TYPE: {BASE64_ENCODED_CREDENTIALS_TYPE}
```

* The `BASE64_ENCODED_CONFIG_JSON` configuration contains credentials and request parameters. 
* The `BASE64_ENCODED_CREDENTIALS_TYPE` assumes one of the following values:  `OAuth`, `BasicAuth`, `Certificate` (not supported in the Director), `NoAuth`.
​

> An example `CONFIGURATION` for APIs secured with OAuth looks as follows:

Reviewer: ```suggestion
The example `CONFIGURATION` for APIs secured with OAuth looks as follows:
```

Reviewer: Also: do we need examples for other authorization types?

Author: I think it should be "an", but the other one is a good question. I'll check.

Author: Yeah, I'll add an example for the `BasicAuth` method, too. Thanks.

```json
{
    "credentials": {
        "clientId": "{OAUTH_CLIENT_ID}",
        "clientSecret": "{OAUTH_CLIENT_SECRET}",
        "requestParameters": {},
        "tokenUrl": "{OAUTH_TOKEN_URL}"
    },
    "csrfConfig": {
        "tokenUrl": "{CSRF_TOKEN_URL}"
    },
    "requestParameters": {
        "headers": {
            "Content-Type": ["application/json"]
        },
        "queryParameters": {
            "limit": ["50"]
        }
    }
}
```

> **NOTE:** All APIs defined in a single Secret use the same configuration - the same credentials, CSRF tokens, and request parameters.

### Calling the Application Gateway
​
The Secret name and the API name are specified as path variables in the following format:

```
http://my-namespace-application-gateway:8080/secret/{SECRET_NAME}/api/{API_NAME}/{CUSTOM_PATH}
```
​

> In order for the Application Gateway to properly read Secrets, they must exist in the same Namespace as the Gateway does.

Reviewer: ```suggestion
In order for the Application Gateway to properly read Secrets, they must exist in the same Namespace as the Gateway.
```
